xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +142,336 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
+		}},
+	})
+	cluster1 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster1Name,
+		ServiceName: newEdsService1Name,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	cluster2 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster2Name,
+		ServiceName: newEdsService2Name,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	t.Log(cluster1, cluster2, ep1, ep2) // XXX
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		Type:        e2e.ClusterTypeAggregate,
+		ClusterName: "envoy.clusters.aggregate",
+		ChildNames:  []string{newCluster1Name, newCluster2Name},
+	})
+	route := e2e.DefaultRouteConfig("new_route", "test.server", newCluster1Name)
+	hashPolicy := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_FilterState_{
+			FilterState: &v3routepb.RouteAction_HashPolicy_FilterState{
+				Key: "io.grpc.channel_id",
+			},
+		},
+	}
+	action := route.VirtualHosts[0].Routes[0].Action.(*v3routepb.Route_Route)
+	action.Route.HashPolicy = []*v3routepb.RouteAction_HashPolicy{&hashPolicy}
+	action.Route.ClusterSpecifier = &v3routepb.RouteAction_Cluster{Cluster: "envoy.clusters.aggregate"}
+	listener := e2e.DefaultClientListener("test.server", route.Name)
+
+	err := xdsServer.Update(context.Background(), e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Clusters:  []*v3clusterpb.Cluster{cluster},
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("failed to update server: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("failed to create client: %s", err)
+	}
+	conn.Connect() // force triggering resource fetch. This isn't great, see comment below.
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Wait to give time to the LB policy to request the child clusters.
+	// This is because go-control-plane, which we rely on for the tests,
+	// refuses to send a snapshot when the client does not request exactly
+	// the resources it contains.
+	// IMO it would be much better if the management server only sent resources
+	// requested by the client, so that we can set it from the beginning and
+	// avoid having to control and time the management server to mach the client
+	// requests.
+	time.Sleep(1 * time.Second)
+
+	err = xdsServer.Update(context.Background(), e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Endpoints: []*endpointv3.ClusterLoadAssignment{ep1, ep2},
+		Clusters:  []*v3clusterpb.Cluster{cluster, cluster1, cluster2},
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("failed to update xDS resources: %v", err)
+	}
+
+	// Here we use the hostname field in the response proto to verify which
```

Comment:
Is it possible to the Peer call option to figure out where the request lands instead of relying on the HostName field in the response? Would that work better?

Reply:
It works equally well and removes some clutter, so I took your suggestion.